File: 2-ad-hoc/1246.cs
using System;
using System.Collections;

class Car{
	public string plate;
	public int start,end;

	public Car(string plate,int start,int end){
		this.plate=plate;
		this.start=start;
		this.end=end;
	}
	public void display(){
		Console.Write(plate+" "+start+" "+end+"|");
	}
}
class Node{
	public Node next;
	public Node prev;

	public Car automotive;

	public Node(Car new_car,Node prev){
		this.automotive=new_car;
		this.next=null;
		this.prev=prev;
	}
}
class Head_ll{
	public Node head;
	public Node first;
	public Node last;

	public Head_ll(){
		this.head=new Node(null,null);
		this.first=head;
	}
	public void insert(Car new_car){
		if(head.automotive==null){
			head.automotive=new_car;
			head.next=new Node(null,head);
			last=head;
			head=head.next;
			return;
		}
		if(head.prev==null){
			head.prev=new Node(new_car,null);
			head.prev.next=head;
			head=head.prev;
			first=head;
			return;
		}

		Node temp=head;
		head=head.prev;
		head.next=new Node(new_car,head);
		head=head.next;
		head.next=temp;
		temp.prev=head;
	}
	public void delete(){
		if(head.prev==null){
			head.automotive=null;
			if(head.next!=null&&head.next.automotive!=null){
				head=head.next;
				head.prev=null;
			}
			first=head;
			return;
		}
		if(head.automotive==null){
			head=head.prev;
			head.automotive=null;
			head.next=null;
			last=head.prev;
			return;
		}
		Node temp=head.next;
		head=head.prev;
		head.next=temp;
		temp.prev=head;
		if(temp.next==null||temp.next.automotive==null)
			last=temp.prev;
	}
	public void move_start(){
		head=first;
	}
	public void move_end(){
		head=last;
	}
	public void display_left(){
		Node temp=first;
		while(temp.automotive!=null){
			temp.automotive.display();
			temp=temp.next;
		}
		Console.Write("\n");
	}
	public void display_right(){
		Node temp=last;
		while(temp!=null&&temp.automotive!=null){
			temp.automotive.display();
			temp=temp.prev;
		}
		Console.WriteLine();
	}
	public bool empty(){
		return (first==last&&first.automotive==null);
	}
}

class Question{
	static void let_parking(Head_ll refers,string plate_car){
		refers.move_start();
		while(refers.head.automotive!=null){
			if(refers.head.automotive.plate==plate_car){
				refers.delete();
				return;
			}
			refers.head=refers.head.next;
		}
	}
	static bool enter_parking(Head_ll refers,int length_park,string plate_car,int length_car){
		refers.move_start();
		if(length_car>length_park)
			return false;
		if(refers.head.automotive==null||refers.head.automotive.start>=length_car){
			refers.insert(new Car(plate_car,0,length_car-1));
			return true;
		}
		int prev_end=refers.head.automotive.end;
		refers.head=refers.head.next;
		while(refers.head.automotive!=null){
			Car temp=refers.head.automotive;
			if(temp.start-prev_end>length_car){
				refers.insert(new Car(plate_car,prev_end+1,prev_end+length_car));
				return true;
			}
			prev_end=temp.end;
			refers.head=refers.head.next;
		}
		if(length_park-prev_end>length_car){
			refers.insert(new Car(plate_car,prev_end+1,prev_end+length_car));
			return true;
		}
		return false;
	}
	static void Main(){
		String[] temp;
		String EOF;
		while((EOF=Console.ReadLine())!=null){
			temp=EOF.Split();
			int length=int.Parse(temp[0]),quotes=int.Parse(temp[1]),tot=0;
			Hashtable desk=new Hashtable();
			Head_ll parking=new Head_ll();
			for(int i=0;i<quotes;++i){
				temp=Console.ReadLine().Split();
				if(temp.Length<3&&desk.ContainsKey(temp[1])){
					let_parking(parking,temp[1]);
					desk.Remove(temp[1]);
				}
				if(temp.Length==3&&int.TryParse(temp[2],out int ok)&&enter_parking(parking,length,temp[1],ok)){
					if(!desk.ContainsKey(temp[1]))
						desk.Add(temp[1],true);
					tot+=10;
				}
			}
			Console.WriteLine(tot);
		}
	}
}

File: 2-ad-hoc/1246.py
import copy

class Car:
    def __init__(self,ID,init,end)->None:
        self.ID=ID;
        self.init=init;
        self.end=end;
        return

    def display(self):
        print("{} {} {}|".format(self.ID,self.init,self.end),end='')

class Node:
    def __init__(self,new_car,prev):
        self.value=copy.deepcopy(new_car)
        self.next=None
        self.prev=prev

class Head_ll:
    def __init__(self):
        self.head=Node(None,None)
        self.first=self.head
        self.last=self.head

    def insert(self,new_value):
        if self.head.value==None:
            self.head.value=new_value
            self.head.next=Node(None,self.head)
            self.head=self.head.next
            self.last=self.head
            return

        if self.head.prev==None:
            self.head.prev=Node(new_value,None)
            self.head.prev.next=self.head
            self.head=self.head.prev
            self.first=self.head
            return

        temp=self.head
        self.head=self.head.prev
        self.head.next=Node(new_value,self.head)

        self.head=self.head.next
        self.head.next=temp
        temp.prev=self.head

    def delete(self):
        if self.head.prev==None:
            self.head.value=None;
            if self.head.next!=None:
                self.head=self.head.next
                self.head.prev=None
            self.first=self.head
            return
        if self.head.value==None:
            self.head=self.head.prev
            self.head.value=None
            self.head.next=None
            self.last=self.head
            return

        temp=self.head.next
        self.head=self.head.prev
        self.head.next=temp
        temp.prev=self.head

    def move_left(self):
        self.head=self.first
    def move_right(self):
        self.head=self.last

    def forward(self,amount_steps):
        for i in range(amount_steps):
            if(self.head.next==None or self.head.next.value==None):
                return;
            self.head=self.head.next
    def back(self,amount_steps):
        for i in range(amount_steps):
            if(self.head.prev==None):
                return
            self.head=self.head.prev

    def display_left(self):
        temp=copy.deepcopy(self)
        while(temp.head.value!=None):
            temp.head.value.display()
            temp.head=temp.head.next
        print()

    def display_right(self):
        temp=copy.deepcopy(self)
        while(temp.head!=None and temp.head.value!=None):
            temp.head.value.display()
            temp.head=temp.head.prev
        print()

def left_parking(parking,ID_car):
    parking.move_left();
    while(parking.head.value!=None and parking.head!=None):
        if parking.head.value.ID==ID_car:
            parking.delete()
            return
        parking.head=parking.head.next

def insert_parking(parking,length_park,ID_car,length_car):
    parking.move_left()
    if length_car>length_park:
        return 0
    if parking.head.value==None or parking.head.value.init>=length_car:
        parking.insert(Car(ID_car,0,length_car-1))
        return 1
    prev_end=parking.head.value.end
    parking.head=parking.head.next
    while(parking.head!=None and parking.head.value!=None):
        curr_car=parking.head.value
        if curr_car.init-prev_end>length_car:
            parking.insert(Car(ID_car,prev_end+1,prev_end+length_car))
            return 1
        prev_end=parking.head.value.end
        parking.head=parking.head.next
    if length_park-prev_end>length_car:
        parking.insert(Car(ID_car,prev_end+1,prev_end+length_car))
        return 1
    return 0



"""
tst=Head_ll()

for i in range(10):
    ID=chr(i%26+65)+chr((i+1)%26+65)+chr((i+2)%26+65)
    temp=Car(ID,i,i)
    tst.insert(temp)

tst.move_left()
tst.head=tst.head.next.next
tst.insert(Car("ALTF4",0,0))
tst.move_right()
tst.delete()
tst.move_left()
tst.display_left()
"""

while 1:
    try:
        [length,quotes]=input().split()
    except:
        break
    length=int(length)
    quotes=int(quotes)
    parking=Head_ll()
    tot=0

    while(quotes):
        inp=input().split()
        if len(inp)<3:
            left_parking(parking,inp[1])
        if len(inp)==3 and insert_parking(parking,length,inp[1],int(inp[2])):
            tot+=10

        quotes-=1
    print(tot)
